File: app/src/main/java/com/bapplications/maplemobile/opengl/utils/Range.java
package com.bapplications.maplemobile.opengl.utils;

public class Range<T extends Comparable<T>> {

    private T upper;
    private T lower;

    public Range(T first, T second) {
        if (first.compareTo(second) <= 0) {
            lower = first;
            upper = second;
        } else {
            lower = second;
            upper = first;
        }
    }

    public T getLower() {
        return lower;
    }

    public T getUpper() {
        return upper;
    }

    public boolean intersect(Range<T> v) {
        return contains(v.lower) || contains(v.upper) || v.contains(lower) || v.contains(upper);
    }

    public boolean contains(T v) {
        return v.compareTo(lower) >= 0 && v.compareTo(upper) < 0;
    }

    public boolean isDot() {
        return getLower().equals(getUpper());
    }
}

File: app/src/main/java/com/bapplications/maplemobile/utils/Range.kt
package com.bapplications.maplemobile.utils

class Range (first: Float, second: Float) {
    var upper: Float = 0f
    var lower: Float = 0f

    fun size(): Float = upper - lower

    fun intersect(v: Range): Boolean = contains(v.lower) || contains(v.upper) ||
            v.contains(lower) || v.contains(upper)

    operator fun contains(v: Float): Boolean = v > lower && v < upper

    operator fun compareTo(x: Float): Int {
        if(x < lower) {
            return 1
        } else if( upper < x) {
            return -1
        }
        return 0
    }

    val center : Float
        get() = lower + (upper - lower) / 2

    val isDot: Boolean
        get() = lower == upper

    init {
        if (first <= second) {
            lower = first
            upper = second
        } else {
            lower = second
            upper = first
        }
    }

    override fun toString(): String {
        return "[$lower, $upper]"
    }
}


